addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,442 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kapppkgiv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/packaging/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	"github.com/vmware-tanzu/tanzu-framework/addons/predicates"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	tkrconstants "github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkr/pkg/constants"
+)
+
+const kappCtrlPkgPrefix = "kapp-controller"
+
+type ClusterRole int
+
+const (
+	clusterRoleManagement ClusterRole = iota
+	clusterRoleWorkload
+)
+
+// PackageInstallStatusReconciler contains the reconciler information for PackageInstallStatus controller
+type PackageInstallStatusReconciler struct {
+	Ctx        context.Context
+	Log        logr.Logger
+	Client     client.Client
+	Scheme     *runtime.Scheme
+	controller controller.Controller
+	Tracker    *util.RemoteObjectTracker
+	// aggregatedAPIResourcesClient is used when it is required to directly read from the server and not to use object caches.
+	aggregatedAPIResourcesClient client.Client
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses/status,verbs=get;update;patch
+
+// Reconcile performs the reconciliation action for the controller; which is reflecting the reconciliation status of each core/additional package
+// into corresponding ClusterBootstrap resource of the cluster. When the cluster is getting deleted, watches for the remote cluster will be removed
+func (r *PackageInstallStatusReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.Log.WithValues(constants.ClusterNamespaceLogKey, req.Namespace, constants.ClusterNameLogKey, req.Name)
+
+	var (
+		c           client.Client
```

Comment:
suggestion: could we rename `c` to something else that makes people easily understand what it would be used for ? Since we have `r.Client`, `c`, `aggregatedAPIResourcesClient` and several reassignments, the usage of them are intertwined. That might cause confusion.

Reply:
as @blc1996 suggested, I'll rename to clusterClient if no objection from you?

Reply:
sounds good to me.